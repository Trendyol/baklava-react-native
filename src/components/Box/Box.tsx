import { createBox } from '@ergenekonyigit/restyle';
import { Theme } from '../../theme';

const Box = createBox<Theme>();

export default Box;
