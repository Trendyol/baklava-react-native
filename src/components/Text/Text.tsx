import { createText } from '@shopify/restyle';
import { Theme } from '../../theme';

const Text = createText<Theme>();

export default Text;
