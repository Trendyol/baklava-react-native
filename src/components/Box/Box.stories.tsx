import { ComponentStory } from '@storybook/react-native';
import React from 'react';
import Text from '../Text/Text';
import Box from './Box';

export default {
  title: 'Box',
  component: Box,
};

type BoxStory = ComponentStory<typeof Box>;

export const Basic: BoxStory = args => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Box
    </Text>
    <Box {...args} />
  </>
);

Basic.args = {
  backgroundColor: 'primaryKey',
  width: 200,
  height: 200,
};
