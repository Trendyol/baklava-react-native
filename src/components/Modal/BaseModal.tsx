import Box, { BoxProps } from '../Box/Box';
import { Animated, ScrollView, StyleSheet } from 'react-native';
import { Portal } from '@gorhom/portal';
import React, { PropsWithChildren } from 'react';
import theme from '../../theme';

export type A11yProps = {
  testID?: string;
  accessibilityLabel?: string;
  accessible?: boolean;
};

export type BaseModalProps = A11yProps &
  PropsWithChildren<{
    modalKey?: string;
    visible: boolean;
    closable?: boolean;
    closeButtonAction?: () => void;
    duration?: number;
    justifyContent?: 'flex-start' | 'flex-end' | 'center';
    wrongPressThresholdMs?: number;
    borderBottomStartRadius?: BoxProps['borderBottomStartRadius'];
    borderBottomEndRadius?: BoxProps['borderBottomEndRadius'];
    p?: BoxProps['p'];
    actions?: React.ReactNode;
    header?: React.ReactNode;
  }>;

const BaseModal = ({
  modalKey,
  justifyContent,
  p = 'l',
  accessible,
  accessibilityLabel,
  testID,
  children,
  header,
  actions,
  visible,
  duration = 200,
  closable = true,
  wrongPressThresholdMs = 1000, // 1 second
  closeButtonAction,
  borderBottomStartRadius,
  borderBottomEndRadius,
}: BaseModalProps) => {
  const anim = React.useRef(new Animated.Value(0)).current;
  const preventWrongPressToClose = React.useRef(Date.now());
  React.useEffect(() => {
    if (visible) {
      preventWrongPressToClose.current = Date.now();
      Animated.timing(anim, {
        toValue: 0.7,
        duration,
        useNativeDriver: true,
      }).start();
    } else {
      anim.setValue(0);
    }
  }, [anim, duration, visible]);

  const close = React.useCallback(() => {
    const threshold = preventWrongPressToClose.current + wrongPressThresholdMs;
    if (closable && threshold < Date.now()) {
      closeButtonAction?.();
    }
  }, [closable, closeButtonAction, wrongPressThresholdMs]);

  const testProps = React.useMemo(() => {
    return {
      accessible: accessible,
      accessibilityLabel: accessibilityLabel ?? testID,
      testID: testID ?? accessibilityLabel,
    } as A11yProps;
  }, [testID, accessibilityLabel, accessible]);

  if (!visible || !children) {
    return null;
  }

  return (
    <Portal name={modalKey ?? 'base-modal'}>
      <Box
        position="absolute"
        justifyContent={justifyContent}
        p={p}
        top={0}
        bottom={0}
        left={0}
        right={0}
        testID={`${testProps.testID}-wrapper`}
        accessibilityLabel={`${testProps.accessibilityLabel}-wrapper`}>
        <Animated.View
          testID={`${testProps.testID}-overlay`}
          accessibilityLabel={`${testProps.testID}-overlay`}
          style={[styles.overlay, { opacity: anim }]}
          onStartShouldSetResponder={() => true}
          onResponderRelease={close}
        />
        <Box
          {...testProps}
          bg="white"
          p="m"
          borderRadius={'l'}
          borderBottomStartRadius={borderBottomStartRadius}
          borderBottomEndRadius={borderBottomEndRadius}
          style={styles.card}>
          {header}
          <ScrollView
            testID={`${testProps.testID}-body`}
            accessibilityLabel={`${testProps.accessibilityLabel}-body`}
            bounces={false}
            showsVerticalScrollIndicator>
            {children}
          </ScrollView>
          {actions}
        </Box>
      </Box>
    </Portal>
  );
};

const styles = StyleSheet.create({
  card: {
    maxHeight: '90%',
  },
  overlay: {
    position: 'absolute',
    backgroundColor: theme.colors.neutralDarker,
    opacity: 0.7,
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
});

export default BaseModal;
