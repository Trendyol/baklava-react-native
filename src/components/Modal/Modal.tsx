import { Portal } from '@gorhom/portal';
import React, { PropsWithChildren } from 'react';
import {
  Animated,
  ScrollView,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import theme from '../../theme';
import Box from '../Box/Box';
import Button, { ButtonProps } from '../Button/Button';
import Icon from '../Icon/Icon';
import Text from '../Text/Text';

type A11yProps = {
  testID?: string;
  accessibilityLabel?: string;
  accessible?: boolean;
};

export type ModalProps = A11yProps &
  PropsWithChildren<{
    visible: boolean;
    title?: string;
    closable?: boolean;
    modalKey?: string;
    closeButtonAction?: () => void;
    actionButtonProps?: ButtonProps;
    secondActionButtonProps?: ButtonProps;
    duration?: number;
    justifyContent?: 'flex-start' | 'flex-end' | 'center';
    wrongPressThresholdMs?: number;
  }>;

const Modal = ({
  visible,
  title,
  closable = true,
  modalKey,
  closeButtonAction,
  actionButtonProps,
  secondActionButtonProps,
  duration = 200,
  justifyContent = 'center',
  wrongPressThresholdMs = 1000, // 1 second
  testID,
  accessibilityLabel,
  accessible = true,
  children,
}: ModalProps) => {
  const anim = React.useRef(new Animated.Value(0)).current;
  const preventWrongPressToClose = React.useRef(Date.now());

  React.useEffect(() => {
    if (visible) {
      preventWrongPressToClose.current = Date.now();
      Animated.timing(anim, {
        toValue: 0.7,
        duration,
        useNativeDriver: true,
      }).start();
    } else {
      anim.setValue(0);
    }
  }, [anim, duration, visible]);

  const testProps = React.useMemo(() => {
    return {
      accessible: accessible,
      accessibilityLabel: accessibilityLabel ?? testID,
      testID: testID ?? accessibilityLabel,
    } as A11yProps;
  }, [testID, accessibilityLabel, accessible]);

  const close = React.useCallback(() => {
    const threshold = preventWrongPressToClose.current + wrongPressThresholdMs;
    if (closable && threshold < Date.now()) {
      closeButtonAction?.();
    }
  }, [closable, closeButtonAction, wrongPressThresholdMs]);

  const header = React.useMemo(() => {
    if (!title && !closable) {
      return null;
    }

    return (
      <Box
        flexDirection="row"
        justifyContent="space-between"
        alignItems="flex-start"
        mb="m"
        testID={`${testProps.testID}-header`}
        accessibilityLabel={`${testProps.accessibilityLabel}-header`}>
        <Box flex={1}>
          <Text
            variant="subtitle2Medium"
            testID={`${testProps.testID}-title`}
            numberOfLines={2}>
            {title}
          </Text>
        </Box>
        {closable && (
          <TouchableOpacity
            testID={`${testProps.testID}-close`}
            accessibilityLabel={`${testProps.accessibilityLabel}-close`}
            hitSlop={{ left: 0, bottom: 0, top: 12, right: 12 }}
            onPress={close}>
            <Box p="4xs" pl="2xs" pb="2xs">
              <Icon name="close" color="neutralDarker" size="xs" />
            </Box>
          </TouchableOpacity>
        )}
      </Box>
    );
  }, [closable, close, testProps.accessibilityLabel, testProps.testID, title]);

  const actions = React.useMemo(() => {
    if (!actionButtonProps && !secondActionButtonProps) {
      return null;
    }

    return (
      <Box
        testID={`${testProps.testID}-actions-area`}
        accessibilityLabel={`${testProps.accessibilityLabel}-actions-area`}>
        {actionButtonProps && (
          <Button
            mt="m"
            filled
            size="l"
            variant="primary"
            testID={`${testProps.testID}-action`}
            accessibilityLabel={`${testProps.accessibilityLabel}-action`}
            {...actionButtonProps}
          />
        )}
        {secondActionButtonProps && (
          <Button
            mt="m"
            filled
            size="l"
            variant="secondary"
            testID={`${testProps.testID}-second-action`}
            accessibilityLabel={`${testProps.accessibilityLabel}-second-action`}
            {...secondActionButtonProps}
          />
        )}
      </Box>
    );
  }, [
    actionButtonProps,
    secondActionButtonProps,
    testProps.accessibilityLabel,
    testProps.testID,
  ]);

  if (!visible || !children) {
    return null;
  }

  return (
    <Portal name={modalKey ?? 'modal'}>
      <Box
        position="absolute"
        justifyContent={justifyContent}
        p="l"
        top={0}
        bottom={0}
        left={0}
        right={0}
        testID={`${testProps.testID}-wrapper`}
        accessibilityLabel={`${testProps.accessibilityLabel}-wrapper`}>
        <Animated.View
          testID={`${testProps.testID}-overlay`}
          accessibilityLabel={`${testProps.testID}-overlay`}
          style={[styles.overlay, { opacity: anim }]}
          onStartShouldSetResponder={() => true}
          onResponderRelease={close}
        />
        <Box
          {...testProps}
          bg="white"
          p="m"
          borderRadius="l"
          style={styles.card}>
          {header}
          <ScrollView
            testID={`${testProps.testID}-body`}
            accessibilityLabel={`${testProps.accessibilityLabel}-body`}
            bounces={false}
            showsVerticalScrollIndicator>
            {children}
          </ScrollView>
          {actions}
        </Box>
      </Box>
    </Portal>
  );
};

const styles = StyleSheet.create({
  card: {
    maxHeight: '90%',
  },
  overlay: {
    position: 'absolute',
    backgroundColor: theme.colors.neutralDarker,
    opacity: 0.7,
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
});

export default Modal;
