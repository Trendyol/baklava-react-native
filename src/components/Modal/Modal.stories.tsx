import { ComponentMeta, ComponentStory } from '@storybook/react-native';
import React from 'react';
import Box from '../Box/Box';
import Button from '../Button/Button';
import Text from '../Text/Text';
import Modal from './Modal';
import Image from '../Image/Image';
import { useWindowDimensions } from 'react-native';

const ModalMeta: ComponentMeta<typeof Modal> = {
  title: 'Modal',
  component: Modal,
  parameters: {
    isFullScreen: true,
    layout: 'centered',
    showPanel: false,
  },
};

export default ModalMeta;

type ModalStory = ComponentStory<typeof Modal>;

export const Basic: ModalStory = () => {
  const dims = useWindowDimensions();
  const [modal1, showModal1] = React.useState(false);
  const [modal2, showModal2] = React.useState(false);
  const [modal3, showModal3] = React.useState(false);
  const [modal4, showModal4] = React.useState(false);
  const [modal5, showModal5] = React.useState(false);
  const [modal6, showModal6] = React.useState(false);

  return (
    <>
      <Box borderRadius="l" p="m">
        <Text variant="subtitle1Bold" mb="l">
          Modal Examples
        </Text>
        <Button mb="m" filled label="Basic" onPress={() => showModal1(true)} />
        <Button
          mb="m"
          filled
          label="Non-Closable"
          onPress={() => showModal2(true)}
        />
        <Button
          mb="m"
          filled
          label="Actions Button"
          onPress={() => showModal3(true)}
        />
        <Button
          mb="m"
          filled
          label="Long Content"
          onPress={() => showModal4(true)}
        />
        <Button
          mb="m"
          filled
          label="Bottom Sticky"
          onPress={() => showModal5(true)}
        />
        <Button
          mb="m"
          filled
          label="Image Content"
          onPress={() => showModal6(true)}
        />
      </Box>
      <Modal
        modalKey="basic"
        visible={modal1}
        closable={true}
        title="Lorem ipsum dolor"
        closeButtonAction={() => showModal1(false)}>
        <Text>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Viverra id arcu convallis facilisi urna proin in.
          Sit sed velit elementum mattis laoreet nulla vitae at. Aenean justo
          nibh enim imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa.
        </Text>
      </Modal>
      <Modal
        modalKey="non-closable"
        visible={modal2}
        closable={false}
        title="Lorem ipsum dolor"
        closeButtonAction={() => showModal2(false)}>
        <Text>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Viverra id arcu convallis facilisi urna proin in.
          Sit sed velit elementum mattis laoreet nulla vitae at. Aenean justo
          nibh enim imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa.
        </Text>
      </Modal>
      <Modal
        modalKey="with-action"
        visible={modal3}
        closable={true}
        title="Lorem ipsum dolor"
        closeButtonAction={() => showModal3(false)}
        actionButtonProps={{
          label: 'Primary Action',
          onPress: () => {},
        }}
        secondActionButtonProps={{
          label: 'Secondary Action',
          onPress: () => {},
        }}>
        <Text>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Viverra id arcu convallis facilisi urna proin in.
          Sit sed velit elementum mattis laoreet nulla vitae at. Aenean justo
          nibh enim imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa.
        </Text>
      </Modal>
      <Modal
        modalKey="long-content"
        visible={modal4}
        closable={true}
        title="Lorem ipsum dolor"
        closeButtonAction={() => showModal4(false)}
        actionButtonProps={{
          label: 'Primary Action',
          onPress: () => {},
        }}
        secondActionButtonProps={{
          label: 'Secondary Action',
          onPress: () => {},
        }}>
        <Text>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Viverra id arcu convallis facilisi urna proin in.
          Sit sed velit elementum mattis laoreet nulla vitae at. Aenean justo
          nibh enim imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          Facilisi arcu accumsan nunc massa. Viverra id arcu convallis facilisi
          urna proin in. Sit sed velit elementum mattis laoreet nulla vitae at.
          Aenean justo nibh enim imperdiet commodo varius eleifend nec vel.
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Lorem ipsum dolor sit amet, consectetur
          adipiscing elit. Facilisi arcu accumsan nunc massa. Viverra id arcu
          convallis facilisi urna proin in. Sit sed velit elementum mattis
          laoreet nulla vitae at. Aenean justo nibh enim imperdiet commodo
          varius eleifend nec vel. Lorem ipsum dolor sit amet, consectetur
          adipiscing elit. Facilisi arcu accumsan nunc massa. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa. Viverra id arcu convallis facilisi urna proin in. Sit sed velit
          elementum mattis laoreet nulla vitae at. Aenean justo nibh enim
          imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor sit amet,
          consectetur adipiscing elit. Facilisi arcu accumsan nunc massa. Lorem
          ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Viverra id arcu convallis facilisi urna proin in.
          Sit sed velit elementum mattis laoreet nulla vitae at. Aenean justo
          nibh enim imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa.
        </Text>
      </Modal>
      <Modal
        modalKey="bottom"
        visible={modal5}
        closable={true}
        justifyContent="flex-end"
        title="Lorem ipsum dolor"
        closeButtonAction={() => showModal5(false)}
        actionButtonProps={{
          label: 'Primary Action',
          onPress: () => {},
        }}
        secondActionButtonProps={{
          label: 'Secondary Action',
          onPress: () => {},
        }}>
        <Text>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Facilisi arcu
          accumsan nunc massa. Viverra id arcu convallis facilisi urna proin in.
          Sit sed velit elementum mattis laoreet nulla vitae at. Aenean justo
          nibh enim imperdiet commodo varius eleifend nec vel. Lorem ipsum dolor
          sit amet, consectetur adipiscing elit. Facilisi arcu accumsan nunc
          massa.
        </Text>
      </Modal>
      <Modal
        modalKey="image-content"
        visible={modal6}
        closable={true}
        title="Image Preview"
        closeButtonAction={() => showModal6(false)}>
        <Box alignItems="center">
          <Image
            height={dims.height * 0.6}
            width={dims.width - 72}
            resizeMode="cover"
            borderRadius={8}
            source={{
              uri: 'https://cdn.dsmcdn.com/ty1006/product/media/images/prod/PIM/20230927/18/a885ced6-7ac8-4ccf-8b58-108db99ed477/1_org_zoom.jpg',
            }}
          />
        </Box>
      </Modal>
    </>
  );
};
