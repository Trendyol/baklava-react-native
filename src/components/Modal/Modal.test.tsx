import React from 'react';
import { Animated } from 'react-native';
import { fireEvent, render } from '../../test-utils';
import Text from '../Text/Text';
import Modal from './Modal';
import Box from '../Box/Box';

describe('Modal', () => {
  test('should not render Modal when visible false', () => {
    // when
    const { toJSON, queryByTestId } = render(
      <Modal testID="modal" visible={false} />,
    );

    const modalWrapper = queryByTestId('modal-wrapper');

    // then
    expect(modalWrapper).toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });

  test('should not render Modal when visible true without children', () => {
    // when
    const { toJSON, queryByTestId } = render(
      <Modal accessibilityLabel="modal" visible={true} />,
    );

    const modalWrapper = queryByTestId('modal-wrapper');

    // then
    expect(modalWrapper).toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render Modal when visible true with children', () => {
    // when
    const { toJSON, queryByTestId } = render(
      <Modal testID="modal" visible={true} title="test title">
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const modalWrapper = queryByTestId('modal-wrapper');
    const title = queryByTestId('modal-title');
    const content = queryByTestId('lorem-text');

    // then
    expect(modalWrapper).not.toBeNull();
    expect(title?.props.children).toBe('test title');
    expect(content).not.toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render Modal without header', () => {
    // when
    const { toJSON, queryByTestId } = render(
      <Modal testID="modal" visible={true} closable={false}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const modalHeader = queryByTestId('modal-header');
    const title = queryByTestId('modal-title');
    const content = queryByTestId('lorem-text');

    // then
    expect(modalHeader).toBeNull();
    expect(title).toBeNull();
    expect(content).not.toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });

  test('should show close button', () => {
    jest.useFakeTimers('modern').setSystemTime(17);
    const mockCloseButtonAction = jest.fn();
    // when
    const { getByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={true}
        title="test title"
        closeButtonAction={mockCloseButtonAction}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const closeBtn = getByTestId('modal-close');
    jest.setSystemTime(1018);
    fireEvent.press(closeBtn);

    // then
    expect(mockCloseButtonAction).toBeCalled();
    jest.useRealTimers();
  });

  test('should not show close button', () => {
    const mockCloseButtonAction = jest.fn();
    // when
    const { queryByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={false}
        title="test title"
        closeButtonAction={mockCloseButtonAction}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const closeBtn = queryByTestId('modal-close');

    // then
    expect(closeBtn).toBeNull();
  });

  test('overlay should close', () => {
    jest.useFakeTimers('modern').setSystemTime(17);
    const mockCloseButtonAction = jest.fn();
    // when
    const { getByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={true}
        wrongPressThresholdMs={5000}
        title="test title"
        closeButtonAction={mockCloseButtonAction}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const overlay = getByTestId('modal-overlay');
    jest.setSystemTime(5018);
    fireEvent(overlay, 'onResponderRelease');

    // then
    expect(mockCloseButtonAction).toBeCalled();
    jest.useRealTimers();
  });

  test('overlay should not close when closable false', () => {
    const mockCloseButtonAction = jest.fn();
    // when
    const { getByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={false}
        title="test title"
        closeButtonAction={mockCloseButtonAction}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const overlay = getByTestId('modal-overlay');
    fireEvent(overlay, 'onResponderRelease');

    // then
    expect(mockCloseButtonAction).not.toBeCalled();
  });

  test('overlay should fade animation', () => {
    const mockStart = jest.fn();
    const mockTiming = jest.fn().mockReturnValue({ start: mockStart });
    jest.spyOn(Animated, 'timing').mockImplementation(mockTiming);
    // when
    render(
      <Modal testID="modal" visible={true} closable={false} title="test title">
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    // then
    expect(mockTiming).toBeCalled();
    expect(mockStart).toBeCalled();
  });

  test('should not show actions area', () => {
    // when
    const { queryByTestId } = render(
      <Modal testID="modal" visible={true} closable={true} title="test title">
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const actionsArea = queryByTestId('modal-actions-area');

    // then
    expect(actionsArea).toBeNull();
  });

  test('should show action button', () => {
    const mockOnPress = jest.fn();
    // when
    const { getByTestId, getByText, queryByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={true}
        title="test title"
        actionButtonProps={{
          label: 'Test Action Button',
          onPress: mockOnPress,
        }}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const actionsArea = getByTestId('modal-actions-area');
    const actionBtn = getByTestId('modal-action');
    const secondActionBtn = queryByTestId('modal-second-action');
    const actionBtnLabel = getByText('Test Action Button');
    fireEvent.press(actionBtn);

    // then
    expect(actionsArea).not.toBeNull();
    expect(secondActionBtn).toBeNull();
    expect(actionBtnLabel).not.toBeNull();
    expect(mockOnPress).toBeCalled();
  });

  test('should show second action button', () => {
    const mockOnPress = jest.fn();
    // when
    const { getByTestId, getByText, queryByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={true}
        title="test title"
        secondActionButtonProps={{
          label: 'Test Second Action Button',
          onPress: mockOnPress,
        }}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const actionsArea = getByTestId('modal-actions-area');
    const actionBtn = queryByTestId('modal-action');
    const secondActionBtn = getByTestId('modal-second-action');
    const secondActionBtnLabel = getByText('Test Second Action Button');
    fireEvent.press(secondActionBtn);

    // then
    expect(actionsArea).not.toBeNull();
    expect(actionBtn).toBeNull();
    expect(secondActionBtnLabel).not.toBeNull();
    expect(mockOnPress).toBeCalled();
  });

  test('should show two action buttons', () => {
    const mockFirstOnPress = jest.fn();
    const mockSecondOnPress = jest.fn();
    // when
    const { getByTestId } = render(
      <Modal
        testID="modal"
        visible={true}
        closable={true}
        title="test title"
        actionButtonProps={{
          label: 'Test Action Button',
          onPress: mockFirstOnPress,
        }}
        secondActionButtonProps={{
          label: 'Test Second Action Button',
          onPress: mockSecondOnPress,
        }}>
        <Text testID="lorem-text">Lorem ipsum dolor</Text>
      </Modal>,
    );

    const actionsArea = getByTestId('modal-actions-area');
    const actionBtn = getByTestId('modal-action');
    const secondActionBtn = getByTestId('modal-second-action');
    fireEvent.press(actionBtn);
    fireEvent.press(secondActionBtn);

    // then
    expect(actionsArea).not.toBeNull();
    expect(mockFirstOnPress).toBeCalled();
    expect(mockSecondOnPress).toBeCalled();
  });

  test('should show multiple modals', () => {
    // when
    const { getByTestId } = render(
      <Box>
        <Text testID="">It shows multiple modals</Text>
        <Modal
          modalKey="first"
          testID="first-modal"
          visible={true}
          closable={true}
          title="first modal">
          <Text testID="first-content">first modal content</Text>
        </Modal>
        <Modal
          modalKey="second"
          testID="second-modal"
          visible={true}
          closable={true}
          title="second modal">
          <Text testID="second-content">second modal content</Text>
        </Modal>
      </Box>,
    );

    const firstModalWrapper = getByTestId('first-modal-wrapper');
    const secondModalWrapper = getByTestId('second-modal-wrapper');

    // then
    expect(firstModalWrapper).not.toBeNull();
    expect(secondModalWrapper).not.toBeNull();
  });
});
