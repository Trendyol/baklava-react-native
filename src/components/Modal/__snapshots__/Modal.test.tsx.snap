// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should not render Modal when visible false 1`] = `null`;

exports[`Modal should not render Modal when visible true without children 1`] = `null`;

exports[`Modal should render Modal when visible true with children 1`] = `
<View
  accessibilityLabel="modal-wrapper"
  style={
    Array [
      Object {
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "padding": 20,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  testID="modal-wrapper"
>
  <View
    accessibilityLabel="modal-overlay"
    collapsable={false}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#273142",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="modal-overlay"
  />
  <View
    accessibilityLabel="modal"
    accessible={true}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "padding": 16,
        },
        Object {
          "maxHeight": "90%",
        },
      ]
    }
    testID="modal"
  >
    <View
      accessibilityLabel="modal-header"
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 16,
          },
        ]
      }
      testID="modal-header"
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <Text
          accessibilityLabel="modal-title"
          accessible={true}
          numberOfLines={2}
          style={
            Array [
              Object {
                "color": "#273142",
                "fontFamily": "Rubik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
                "lineHeight": 20,
              },
            ]
          }
          testID="modal-title"
        >
          test title
        </Text>
      </View>
      <View
        accessibilityLabel="modal-close"
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 0,
            "left": 0,
            "right": 12,
            "top": 12,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="modal-close"
      >
        <View
          style={
            Array [
              Object {
                "padding": 2,
                "paddingBottom": 8,
                "paddingLeft": 8,
              },
            ]
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight="14"
            bbWidth="14"
            fill="#273142"
            focusable={false}
            height={14}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 14,
                  "width": 14,
                },
              ]
            }
            testID="icon"
            title="close"
            vbHeight={24}
            vbWidth={24}
            width={14}
            xmlns="http://www.w3.org/2000/svg"
          >
            <RNSVGGroup
              fill={
                Object {
                  "type": 0,
                  "value": 4280758594,
                }
              }
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                clipRule={0}
                d="M.54.54a1.846 1.846 0 0 1 2.612 0L12 9.39 20.848.54a1.846 1.846 0 0 1 2.611 2.612L14.611 12l8.848 8.848a1.846 1.846 0 0 1-2.61 2.611L12 14.611l-8.848 8.848A1.846 1.846 0 0 1 .54 20.85L9.389 12 .541 3.152A1.846 1.846 0 0 1 .54.54Z"
                fill={
                  Object {
                    "type": 0,
                    "value": 4280758594,
                  }
                }
                fillRule={0}
                propList={
                  Array [
                    "fill",
                    "fillRule",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
    <RCTScrollView
      accessibilityLabel="modal-body"
      bounces={false}
      showsVerticalScrollIndicator={true}
      testID="modal-body"
    >
      <View>
        <Text
          accessibilityLabel="lorem-text"
          accessible={true}
          style={
            Array [
              Object {
                "color": "#273142",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
              },
            ]
          }
          testID="lorem-text"
        >
          Lorem ipsum dolor
        </Text>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`Modal should render Modal without header 1`] = `
<View
  accessibilityLabel="modal-wrapper"
  style={
    Array [
      Object {
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "padding": 20,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  testID="modal-wrapper"
>
  <View
    accessibilityLabel="modal-overlay"
    collapsable={false}
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#273142",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="modal-overlay"
  />
  <View
    accessibilityLabel="modal"
    accessible={true}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "padding": 16,
        },
        Object {
          "maxHeight": "90%",
        },
      ]
    }
    testID="modal"
  >
    <RCTScrollView
      accessibilityLabel="modal-body"
      bounces={false}
      showsVerticalScrollIndicator={true}
      testID="modal-body"
    >
      <View>
        <Text
          accessibilityLabel="lorem-text"
          accessible={true}
          style={
            Array [
              Object {
                "color": "#273142",
                "fontFamily": "Rubik-Regular",
                "fontSize": 14,
              },
            ]
          }
          testID="lorem-text"
        >
          Lorem ipsum dolor
        </Text>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
